In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set up for better plots/
plt.style.use('seaborn-v0_8')
sns.set_palette('husl')

# Data Processing

## Dataset 1: India Data Portal - Employment Generated

In [3]:
EmpGen = pd.read_csv('/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/1. Raw/India Data Portal (PC lvl)/Employment Generated.csv')
EmpGen.head()

,_id,id,year,state_name,district_name,block_name,gp_name,reg_hh,reg_pers,del_jobcards_hh,...,emp_offer_hh,emp_offer_pers,emp_avail_hh,emp_avail_pers,emp_avail_tot_persondays,emp_avail_central_persondays,emp_avail_states_persondays,fam_completed_100_days,land_reform_benef_hh,disabled_benef_indiv
0,1,0,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Campbell Bay,376.0,800.0,0.0,...,29.0,35.0,16.0,20.0,381.0,381.0,0.0,0.0,2.0,0.0
1,2,1,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Govindnagar,429.0,813.0,0.0,...,88.0,94.0,62.0,66.0,745.0,745.0,0.0,0.0,0.0,0.0
2,3,2,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Great & Little Nicobar,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,4,3,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Laxmi Nagar,460.0,819.0,0.0,...,30.0,30.0,29.0,29.0,214.0,214.0,0.0,0.0,0.0,0.0
4,5,4,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Chowra Tc,309.0,632.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


### 1.1 Checking what information is stored in the dataset

In [4]:
EmpGen['year'].value_counts()

year
2021-2022    269828
2022-2023    269746
2020-2021    269539
2023-2024    269160
2018-2019    267030
2019-2020    266762
2016-2017    258862
2017-2018    258145
2015-2016    257344
2014-2015    256134
Name: count, dtype: int64

In [3]:
EmpGen.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2642550 entries, 0 to 2642549
Data columns (total 29 columns):
 #   Column                        Dtype  
---  ------                        -----  
 0   _id                           int64  
 1   id                            int64  
 2   year                          object 
 3   state_name                    object 
 4   district_name                 object 
 5   block_name                    object 
 6   gp_name                       object 
 7   reg_hh                        float64
 8   reg_pers                      float64
 9   del_jobcards_hh               float64
 10  del_jobcards_pers             float64
 11  incl_jobcards_hh              float64
 12  incl_jobcards_pers            float64
 13  cumul_hh_jobcards_sc          float64
 14  cumul_hh_jobcards_sts         float64
 15  cumul_hh_jobcards_others      float64
 16  cumul_hh_jobcards_tot         float64
 17  emp_demand_hh                 float64
 18  emp_demand_pers       

### 1.2 Confirming the key of the dataset

In [4]:
EmpGen[['year','state_name','district_name','block_name','gp_name']].value_counts().unique()

array([2, 1])

Seems like there are some 'keys' that actually appear twice instead of once. Let's select the entries for these keys and store them in a new dataset.

In [5]:
EmpGen_dup = EmpGen[EmpGen.duplicated(
    subset=['year','state_name','district_name','block_name','gp_name'],
    keep=False
)]

### 1.3 Dealing with duplicates

Since we only care about specific columns, let's see how the duplicates behave for those specific columns.

In [6]:
cols = ['id',
    'year','state_name','district_name','block_name','gp_name',
    'reg_hh', 'reg_pers',
    'cumul_hh_jobcards_sc',
    'cumul_hh_jobcards_sts',
    'cumul_hh_jobcards_others',
    'cumul_hh_jobcards_tot']

reg_hh - number of registered households <br>
reg_pers - number of registered persons <br>

cumul_hh_jobcards_sc - Cummulative job cards issued for sc households <br>
cumul_hh_jobcards_sts - Cummulative job cards issued for st households <br>
cumul_hh_jobcards_others - Cummulative job cards issued for other households <br>
cumul_hh_jobcards_tot - Cummulative job cards issued for all households

In [7]:
EmpGen_dup[['reg_hh', 'reg_pers', 'cumul_hh_jobcards_sc',
     'cumul_hh_jobcards_sts',
     'cumul_hh_jobcards_others',
     'cumul_hh_jobcards_tot']].head(10)

,reg_hh,reg_pers,cumul_hh_jobcards_sc,cumul_hh_jobcards_sts,cumul_hh_jobcards_others,cumul_hh_jobcards_tot
2607,1235.0,2652.0,380.0,58.0,797.0,1235.0
2608,590.0,1072.0,20.0,5.0,565.0,590.0
14777,85.0,173.0,0.0,85.0,0.0,85.0
14790,0.0,0.0,0.0,0.0,0.0,0.0
16512,14760.0,29180.0,591.0,12967.0,1189.0,14747.0
16513,6.0,8.0,0.0,0.0,0.0,0.0
40595,210.0,545.0,3.0,196.0,11.0,210.0
40596,584.0,1380.0,0.0,518.0,61.0,579.0
40626,0.0,0.0,0.0,0.0,0.0,0.0
40627,424.0,1053.0,1.0,417.0,6.0,424.0


Let's check what these duplicates really look like for some randomly selected GPs.

In [8]:
EmpGen_dup

,_id,id,year,state_name,district_name,block_name,gp_name,reg_hh,reg_pers,del_jobcards_hh,...,emp_offer_hh,emp_offer_pers,emp_avail_hh,emp_avail_pers,emp_avail_tot_persondays,emp_avail_central_persondays,emp_avail_states_persondays,fam_completed_100_days,land_reform_benef_hh,disabled_benef_indiv
2607,2608,2607,2014-2015,Andhra Pradesh,Cuddapah,Chitvel,Chitvel,1235.0,2652.0,0.0,...,253.0,358.0,223.0,294.0,7443.0,7443.0,0.0,11.0,36.0,1.0
2608,2609,2608,2014-2015,Andhra Pradesh,Cuddapah,Chitvel,Chitvel,590.0,1072.0,0.0,...,4.0,4.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
14777,14778,14777,2014-2015,Arunachal Pradesh,West Kameng,Nafra,Mathow,85.0,173.0,0.0,...,76.0,77.0,76.0,77.0,1737.0,1737.0,0.0,0.0,0.0,0.0
14790,14791,14790,2014-2015,Arunachal Pradesh,West Kameng,Nafra,Mathow,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
16512,16513,16512,2014-2015,Assam,Karbi Anglong,Samelangso,No Panchayat,14760.0,29180.0,26.0,...,9593.0,9690.0,9362.0,9395.0,52153.0,52153.0,0.0,0.0,184.0,5.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
2638204,2638205,2638204,2023-2024,Uttarakhand,Tehri Garhwal,Pratapnagar,Oanalgaon,140.0,158.0,1.0,...,109.0,109.0,109.0,109.0,1952.0,1952.0,0.0,0.0,0.0,0.0
2638212,2638213,2638212,2023-2024,Uttarakhand,Tehri Garhwal,Pratapnagar,Pokhri,178.0,194.0,0.0,...,37.0,37.0,29.0,29.0,406.0,406.0,0.0,0.0,0.0,0.0
2638213,2638214,2638213,2023-2024,Uttarakhand,Tehri Garhwal,Pratapnagar,Pokhri,243.0,412.0,0.0,...,202.0,297.0,104.0,151.0,2206.0,2206.0,0.0,0.0,0.0,0.0
2638215,2638216,2638215,2023-2024,Uttarakhand,Tehri Garhwal,Pratapnagar,Pujargaon,170.0,230.0,0.0,...,50.0,55.0,49.0,54.0,735.0,735.0,0.0,0.0,0.0,0.0


In [9]:
display(EmpGen[(EmpGen['district_name'] == 'Cuddapah') & (EmpGen['gp_name'] == 'Chitvel')])
display(EmpGen[(EmpGen['district_name'] == 'West Kameng') & (EmpGen['gp_name'] == 'Mathow')].head(8))
display(EmpGen[(EmpGen['district_name'] == 'Tehri Garhwal') & (EmpGen['gp_name'] == 'Oanalgaon')].head(8))

,_id,id,year,state_name,district_name,block_name,gp_name,reg_hh,reg_pers,del_jobcards_hh,...,emp_offer_hh,emp_offer_pers,emp_avail_hh,emp_avail_pers,emp_avail_tot_persondays,emp_avail_central_persondays,emp_avail_states_persondays,fam_completed_100_days,land_reform_benef_hh,disabled_benef_indiv
2607,2608,2607,2014-2015,Andhra Pradesh,Cuddapah,Chitvel,Chitvel,1235.0,2652.0,0.0,...,253.0,358.0,223.0,294.0,7443.0,7443.0,0.0,11.0,36.0,1.0
2608,2609,2608,2014-2015,Andhra Pradesh,Cuddapah,Chitvel,Chitvel,590.0,1072.0,0.0,...,4.0,4.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
258741,258742,258741,2015-2016,Andhra Pradesh,Cuddapah,Chitvel,Chitvel,655.0,1606.0,590.0,...,124.0,187.0,121.0,166.0,5821.0,5821.0,0.0,19.0,26.0,0.0
258742,258743,258742,2015-2016,Andhra Pradesh,Cuddapah,Chitvel,Chitvel,615.0,1106.0,1.0,...,117.0,145.0,103.0,119.0,3810.0,3810.0,0.0,1.0,16.0,0.0


,_id,id,year,state_name,district_name,block_name,gp_name,reg_hh,reg_pers,del_jobcards_hh,...,emp_offer_hh,emp_offer_pers,emp_avail_hh,emp_avail_pers,emp_avail_tot_persondays,emp_avail_central_persondays,emp_avail_states_persondays,fam_completed_100_days,land_reform_benef_hh,disabled_benef_indiv
14777,14778,14777,2014-2015,Arunachal Pradesh,West Kameng,Nafra,Mathow,85.0,173.0,0.0,...,76.0,77.0,76.0,77.0,1737.0,1737.0,0.0,0.0,0.0,0.0
14790,14791,14790,2014-2015,Arunachal Pradesh,West Kameng,Nafra,Mathow,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
270911,270912,270911,2015-2016,Arunachal Pradesh,West Kameng,Nafra,Mathow,85.0,173.0,0.0,...,57.0,58.0,55.0,55.0,696.0,696.0,0.0,0.0,0.0,0.0
270924,270925,270924,2015-2016,Arunachal Pradesh,West Kameng,Nafra,Mathow,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
529086,529087,529086,2016-2017,Arunachal Pradesh,West Kameng,Nafra,Mathow,83.0,166.0,2.0,...,82.0,91.0,82.0,91.0,2272.0,2272.0,0.0,0.0,0.0,0.0
529098,529099,529098,2016-2017,Arunachal Pradesh,West Kameng,Nafra,Mathow,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
787770,787771,787770,2017-2018,Arunachal Pradesh,West Kameng,Nafra,Mathow,74.0,138.0,9.0,...,76.0,76.0,74.0,74.0,1098.0,1098.0,0.0,0.0,0.0,0.0
787782,787783,787782,2017-2018,Arunachal Pradesh,West Kameng,Nafra,Mathow,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


,_id,id,year,state_name,district_name,block_name,gp_name,reg_hh,reg_pers,del_jobcards_hh,...,emp_offer_hh,emp_offer_pers,emp_avail_hh,emp_avail_pers,emp_avail_tot_persondays,emp_avail_central_persondays,emp_avail_states_persondays,fam_completed_100_days,land_reform_benef_hh,disabled_benef_indiv
251766,251767,251766,2014-2015,Uttarakhand,Tehri Garhwal,Pratapnagar,Oanalgaon,127.0,220.0,0.0,...,103.0,104.0,102.0,103.0,2129.0,2129.0,0.0,0.0,0.0,0.0
251767,251768,251767,2014-2015,Uttarakhand,Tehri Garhwal,Pratapnagar,Oanalgaon,156.0,286.0,0.0,...,116.0,121.0,109.0,114.0,5479.0,5479.0,0.0,0.0,0.0,0.0
509114,509115,509114,2015-2016,Uttarakhand,Tehri Garhwal,Pratapnagar,Oanalgaon,202.0,295.0,0.0,...,198.0,211.0,168.0,178.0,5117.0,5117.0,0.0,1.0,0.0,0.0
509115,509116,509115,2015-2016,Uttarakhand,Tehri Garhwal,Pratapnagar,Oanalgaon,160.0,290.0,0.0,...,112.0,113.0,111.0,112.0,5116.0,5116.0,0.0,0.0,0.0,0.0
767953,767954,767953,2016-2017,Uttarakhand,Tehri Garhwal,Pratapnagar,Oanalgaon,186.0,271.0,16.0,...,176.0,183.0,170.0,177.0,8808.0,8808.0,0.0,0.0,0.0,0.0
767954,767955,767954,2016-2017,Uttarakhand,Tehri Garhwal,Pratapnagar,Oanalgaon,123.0,220.0,40.0,...,122.0,126.0,122.0,126.0,7267.0,7267.0,0.0,0.0,0.0,0.0
1026098,1026099,1026098,2017-2018,Uttarakhand,Tehri Garhwal,Pratapnagar,Oanalgaon,180.0,205.0,14.0,...,166.0,171.0,158.0,161.0,8654.0,8654.0,0.0,0.0,0.0,0.0
1026099,1026100,1026099,2017-2018,Uttarakhand,Tehri Garhwal,Pratapnagar,Oanalgaon,124.0,136.0,1.0,...,121.0,123.0,116.0,118.0,9397.0,9397.0,0.0,0.0,0.0,0.0


These results tell us that there are three types of cases.

Case 1 (Mathow): Real rows accompanied by garbage 'zero' rows

Case 2 (Chitvel): Data Spliting - Two rows with different non-zero values. One looks “complete”, one looks partial, hinting at split reporting

Case 3 (Oanalgaon): Conflicting full values - Both rows have non-zero, substantial values, resulting in true duplication ambiguity

We can remove the zero rows first

In [10]:
EmpGen_clean = EmpGen.copy()
num_cols = ['reg_hh', 'reg_pers',
    'cumul_hh_jobcards_sc',
    'cumul_hh_jobcards_sts',
    'cumul_hh_jobcards_others',
    'cumul_hh_jobcards_tot']
EmpGen_clean = EmpGen_clean[EmpGen_clean[num_cols].sum(axis=1) > 0]

For the rest, we can use the max() function to keep the max of the two values.

Duplicate entries likely arise from administrative reporting inconsistencies. Since variables are cumulative counts, the maximum value across duplicates best approximates the final recorded total while avoiding double counting.

In [11]:
EmpGen_clean = EmpGen_clean.groupby(
    ['year','state_name','district_name','block_name','gp_name'],
    as_index=False
)[num_cols].max()

In [12]:
EmpGen_clean[['year','state_name','district_name','block_name','gp_name']].value_counts().unique()

array([1])

Now all the keys are unique!

### 1.4 Selecting necessarily columns

In [13]:
JobCards = EmpGen_clean[['year','state_name', 'district_name', 'block_name', 'gp_name', 'reg_hh', 'reg_pers', 'cumul_hh_jobcards_sc', 
                   'cumul_hh_jobcards_sts', 'cumul_hh_jobcards_others', 'cumul_hh_jobcards_tot']]
JobCards.head()

,year,state_name,district_name,block_name,gp_name,reg_hh,reg_pers,cumul_hh_jobcards_sc,cumul_hh_jobcards_sts,cumul_hh_jobcards_others,cumul_hh_jobcards_tot
0,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Campbell Bay,376.0,800.0,4.0,2.0,369.0,375.0
1,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Govindnagar,429.0,813.0,0.0,0.0,429.0,429.0
2,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Laxmi Nagar,460.0,819.0,0.0,0.0,460.0,460.0
3,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Chowra Tc,309.0,632.0,0.0,309.0,0.0,309.0
4,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Kamorta Tc,612.0,1344.0,0.0,491.0,121.0,612.0


### 1.5 Internal consistency: do SC + ST + Others add up to Total?

Since `cumul_hh_jobcards_tot` is reported separately from the three caste sub-categories, we can verify whether `sc + sts + others == tot`. Large discrepancies would directly undermine any equity analysis built on these variables.

In [14]:
JobCards['components_sum'] = (
    JobCards['cumul_hh_jobcards_sc'] +
    JobCards['cumul_hh_jobcards_sts'] +
    JobCards['cumul_hh_jobcards_others']
)
JobCards['tot_discrepancy'] = JobCards['cumul_hh_jobcards_tot'] - JobCards['components_sum']

n_inconsistent = (JobCards['tot_discrepancy'].abs() > 0).sum()
print(f"Rows where sc + sts + others != tot: {n_inconsistent} ({100*n_inconsistent/len(JobCards):.2f}%)")
print()
print("Distribution of discrepancy (tot - sum of components):")
print(JobCards['tot_discrepancy'].describe())


Rows where sc + sts + others != tot: 104 (0.00%)

Distribution of discrepancy (tot - sum of components):
count    2.600926e+06
mean    -1.988523e-03
std      6.976244e-01
min     -3.940000e+02
25%      0.000000e+00
50%      0.000000e+00
75%      0.000000e+00
max      0.000000e+00
Name: tot_discrepancy, dtype: float64


Mostly the discrepancies are small (median is 0, mean is around 2), but there are some large outliers (max discrepancy is around 400) which may warrant further investigation.

Let's sort out the rows where the discrepancies are greater than 5% of the total number reported

In [15]:
JobCards['gap_pct'] = (
    (JobCards['tot_discrepancy'] / JobCards['cumul_hh_jobcards_tot']) * 100
)

inconsistent = (JobCards['gap_pct'].abs() > 0.05).sum()
print(f"Rows where inconsistency is greater than 5% : {inconsistent} ({100*inconsistent/len(JobCards):.2f}%)")


Rows where inconsistency is greater than 5% : 104 (0.00%)


A very small % of PCs display discrepancies greater than 5%. Therefore, we can just drop these rows for our analysis.

In [16]:
JobCards_filtered = JobCards[(JobCards['gap_pct'].abs() <= 0.05)].drop(
                                 columns = ['components_sum', 'tot_discrepancy', 'gap_pct']
                             )
JobCards_filtered.head()

,year,state_name,district_name,block_name,gp_name,reg_hh,reg_pers,cumul_hh_jobcards_sc,cumul_hh_jobcards_sts,cumul_hh_jobcards_others,cumul_hh_jobcards_tot
0,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Campbell Bay,376.0,800.0,4.0,2.0,369.0,375.0
1,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Govindnagar,429.0,813.0,0.0,0.0,429.0,429.0
2,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Laxmi Nagar,460.0,819.0,0.0,0.0,460.0,460.0
3,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Chowra Tc,309.0,632.0,0.0,309.0,0.0,309.0
4,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Kamorta Tc,612.0,1344.0,0.0,491.0,121.0,612.0


### 1.6 Monotonicity of cumulative variables over time

Since the job card variables are cumulative counts, they should be non-decreasing for the same GP across years. A year-on-year decline signals an administrative reset, data entry error, or GP boundary change — all of which are important to flag before using these as outcome variables.

In [17]:
gp_key = ['state_name', 'district_name', 'block_name', 'gp_name']
cumul_cols = ['cumul_hh_jobcards_sc', 'cumul_hh_jobcards_sts',
              'cumul_hh_jobcards_others', 'cumul_hh_jobcards_tot']

# Sort by GP then year, compute year-on-year differences within each GP
JobCards_sorted = JobCards_filtered.sort_values(gp_key + ['year'])
diffs = JobCards_sorted.groupby(gp_key)[cumul_cols].diff()

# A violation is any row where any cumulative column decreased
violations = (diffs < 0).any(axis=1)
n_violations = violations.sum()
print(f"GP-year rows with at least one cumulative column decreasing: {n_violations} ({100*n_violations/len(JobCards):.2f}%)")
print()

# Break down by which column is violated most
print("Violations per column:")
print((diffs < 0).sum())


GP-year rows with at least one cumulative column decreasing: 695136 (26.73%)

Violations per column:
cumul_hh_jobcards_sc        443892
cumul_hh_jobcards_sts       235189
cumul_hh_jobcards_others    498068
cumul_hh_jobcards_tot       527857
dtype: int64


In [18]:
# Inspect a sample of violating rows alongside their previous year
violating_idx = JobCards_sorted[violations].index
sample_violations = JobCards_sorted.loc[violating_idx].head(10)[
    gp_key + ['year'] + cumul_cols
]
display(sample_violations)


,state_name,district_name,block_name,gp_name,year,cumul_hh_jobcards_sc,cumul_hh_jobcards_sts,cumul_hh_jobcards_others,cumul_hh_jobcards_tot
754744,Andaman And Nicobar,Nicobars,Campbell Bay,Campbell Bay,2017-2018,4.0,2.0,377.0,383.0
1795918,Andaman And Nicobar,Nicobars,Campbell Bay,Campbell Bay,2021-2022,0.0,0.0,369.0,369.0
2064428,Andaman And Nicobar,Nicobars,Campbell Bay,Campbell Bay,2022-2023,0.0,0.0,367.0,367.0
251498,Andaman And Nicobar,Nicobars,Campbell Bay,Govindnagar,2015-2016,0.0,0.0,428.0,428.0
754745,Andaman And Nicobar,Nicobars,Campbell Bay,Govindnagar,2017-2018,0.0,0.0,430.0,430.0
1795919,Andaman And Nicobar,Nicobars,Campbell Bay,Govindnagar,2021-2022,0.0,0.0,409.0,409.0
2064429,Andaman And Nicobar,Nicobars,Campbell Bay,Govindnagar,2022-2023,0.0,0.0,407.0,407.0
754746,Andaman And Nicobar,Nicobars,Campbell Bay,Laxmi Nagar,2017-2018,0.0,0.0,456.0,456.0
251501,Andaman And Nicobar,Nicobars,Nancowry,Kamorta Tc,2015-2016,0.0,490.0,121.0,611.0
503355,Andaman And Nicobar,Nicobars,Nancowry,Kamorta Tc,2016-2017,0.0,493.0,114.0,607.0


### 1.7 Missing values

Checking for NaN values across all columns in `JobCards`. Even though the zero-row filter in 1.3 removed all-zero rows, it is possible that individual cells are NaN rather than zero — particularly in the caste-disaggregated columns, which are the key equity variables.

In [19]:
missing = JobCards_filtered.isnull().sum()
missing_pct = 100 * missing / len(JobCards_filtered)

missing_summary = pd.DataFrame({'missing_count': missing, 'missing_pct': missing_pct.round(2)})
display(missing_summary[missing_summary['missing_count'] > 0])

print(f"\nTotal rows in JobCards_filtered: {len(JobCards_filtered)}")
print(f"Rows with any missing value: {JobCards_filtered.isnull().any(axis=1).sum()}")


,missing_count,missing_pct



Total rows in JobCards_filtered: 2598530
Rows with any missing value: 0


### 1.8 State × year coverage: GP counts

A pivot table showing the number of GPs observed per state per year. This reveals whether coverage is stable across the panel — sudden drops or gaps in a state-year cell indicate that data was not reported or not collected for that year, which matters for any panel fixed-effects analysis.

In [20]:
coverage = JobCards_filtered.pivot_table(
    index='state_name',
    columns='year',
    values='gp_name',
    aggfunc='count'
)

# Sort years chronologically (they are strings like "2014-2015")
coverage = coverage[sorted(coverage.columns)]

# Highlight missing (NaN) cells and large drops to make gaps visible
coverage.style.background_gradient(cmap='YlGn', axis=1).format('{:.0f}', na_rep='—')


year,2014-2015,2015-2016,2016-2017,2017-2018,2018-2019,2019-2020,2020-2021,2021-2022,2022-2023,2023-2024
state_name,,,,,,,,,,
Andaman And Nicobar,81,81,76,76,76,76,77,77,77,77
Andhra Pradesh,13076,13078,13078,12915,12915,12917,13055,13281,13378,13380
Arunachal Pradesh,1800,1800,1798,1798,1798,1801,1939,2036,2078,2079
Assam,2664,2664,2663,2663,2663,2661,2660,2660,2660,2660
Bihar,8494,8490,8399,8396,8395,8395,8396,8389,8333,8332
Chhattisgarh,10945,10953,10949,10952,10956,11641,11646,11654,11655,11653
Goa,180,184,186,186,186,186,186,188,188,188
Gujarat,13824,13852,13749,13893,13919,13960,14053,14154,14200,14179
Haryana,5981,6025,6047,6044,6075,6109,6174,6184,6187,6188


Let's take a look at how the data cleaning reshaped our data:

In [21]:
print(f"\nTotal rows and columns in Raw Data (EmpGen): {EmpGen.shape}")
print(f"\nTotal rows and columns in Cleaned Data (JobCards_filtered): {JobCards_filtered.shape}")

print(f'{len(EmpGen) - len(JobCards_filtered)} rows ({100*(len(EmpGen) - len(JobCards_filtered))/len(EmpGen):.2f}%) removed during cleaning')


Total rows and columns in Raw Data (EmpGen): (2642550, 29)

Total rows and columns in Cleaned Data (JobCards_filtered): (2598530, 11)
44020 rows (1.67%) removed during cleaning


In [61]:
JobCards_filtered.to_csv('/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/3. Processed/JobCards_filtered.csv', index=False)

## Dataset 2: India Data Portal - Workers

### 2.1 Checking what information is stored in the dataset

In [22]:
Workers = pd.read_csv('/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/1. Raw/India Data Portal (PC lvl)/Category Wise Workers.csv')
Workers.head()

,_id,id,year,state_name,district_name,block_name,gp_name,applied_jobcards,issued_jobcards,registered_workers_scs,registered_workers_sts,registered_workers_others,total_registered_workers,registered_workers_women,active_jobcards,active_workers_scs,active_workers_sts,active_workers_others,total_active_workers,active_workers_women
0,1,0.0,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Chowra Tc,309.0,309.0,0.0,646.0,2.0,648.0,317.0,150.0,0.0,63.0,2.0,65.0,24.0
1,2,1.0,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Kamorta Tc,612.0,612.0,0.0,1255.0,191.0,1446.0,740.0,388.0,0.0,443.0,15.0,458.0,218.0
2,3,2.0,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Katchal Tc,452.0,452.0,0.0,435.0,307.0,742.0,299.0,273.0,0.0,200.0,89.0,289.0,114.0
3,4,3.0,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Nancowry Tc,236.0,236.0,0.0,590.0,1.0,591.0,279.0,177.0,0.0,177.0,1.0,178.0,90.0
4,5,4.0,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Teressa Tc,332.0,332.0,0.0,580.0,1.0,581.0,237.0,257.0,0.0,78.0,1.0,79.0,21.0


In [23]:
Workers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2676899 entries, 0 to 2676898
Data columns (total 20 columns):
 #   Column                     Dtype  
---  ------                     -----  
 0   _id                        int64  
 1   id                         float64
 2   year                       object 
 3   state_name                 object 
 4   district_name              object 
 5   block_name                 object 
 6   gp_name                    object 
 7   applied_jobcards           float64
 8   issued_jobcards            float64
 9   registered_workers_scs     float64
 10  registered_workers_sts     float64
 11  registered_workers_others  float64
 12  total_registered_workers   float64
 13  registered_workers_women   float64
 14  active_jobcards            float64
 15  active_workers_scs         float64
 16  active_workers_sts         float64
 17  active_workers_others      float64
 18  total_active_workers       float64
 19  active_workers_women       float64
dtypes:

In [24]:
display(Workers['year'].unique())
Workers['state_name'].unique()

array(['2014-2015', '2015-2016', '2016-2017', '2017-2018', '2018-2019',
       '2019-2020', '2020-2021', '2021-2022', '2022-2023', '2023-2024'],
      dtype=object)

array(['Andaman And Nicobar', 'Andhra Pradesh', 'Arunachal Pradesh',
       'Assam', 'Bihar', 'Chhattisgarh', 'Dn Haveli And Dd', 'Goa',
       'Gujarat', 'Haryana', 'Himachal Pradesh', 'Jammu And Kashmir',
       'Jharkhand', 'Karnataka', 'Kerala', 'Ladakh', 'Lakshadweep',
       'Madhya Pradesh', 'Maharashtra', 'Manipur', 'Meghalaya', 'Mizoram',
       'Nagaland', 'Odisha', 'Puducherry', 'Punjab', 'Rajasthan',
       'Sikkim', 'Tamil Nadu', 'Telangana', 'Tripura', 'Uttar Pradesh',
       'Uttarakhand', 'West Bengal'], dtype=object)

### 2.2 Confirming the key of the dataset

In [25]:
Workers[['year','state_name','district_name','block_name','gp_name']].value_counts().unique()

array([2, 1])

Since the results contain both 2 and 1, some GP-year combinations appear twice and we need to resolve them before analysis.

In [26]:
Workers_dup = Workers[Workers.duplicated(
    subset=['year','state_name','district_name','block_name','gp_name'],
    keep=False
)]
print(f"Rows involved in duplicates: {len(Workers_dup)} ({100*len(Workers_dup)/len(Workers):.2f}%)")

Rows involved in duplicates: 1038 (0.04%)


### 2.3 Dealing with duplicates

Let's inspect the numeric columns for the duplicated rows to understand what type of duplication we're dealing with — the same three cases can arise here as in Dataset 1 (zero ghost rows, split reporting, or conflicting full values).

In [27]:
worker_numcols = [
    'applied_jobcards', 'issued_jobcards',
    'registered_workers_scs', 'registered_workers_sts',
    'registered_workers_others', 'total_registered_workers',
    'registered_workers_women',
    'active_jobcards',
    'active_workers_scs', 'active_workers_sts',
    'active_workers_others', 'total_active_workers',
    'active_workers_women'
]

Workers_dup[['year','state_name','district_name','block_name','gp_name'] + worker_numcols].head(20)

,year,state_name,district_name,block_name,gp_name,applied_jobcards,issued_jobcards,registered_workers_scs,registered_workers_sts,registered_workers_others,total_registered_workers,registered_workers_women,active_jobcards,active_workers_scs,active_workers_sts,active_workers_others,total_active_workers,active_workers_women
3132,2014-2015,Andhra Pradesh,Cuddapah,Chitvel,Chitvel,590.0,590.0,6.0,3.0,609.0,618.0,371.0,0.0,1.0,3.0,367.0,371.0,239.0
3133,2014-2015,Andhra Pradesh,Cuddapah,Chitvel,Chitvel,1235.0,1235.0,485.0,45.0,123.0,653.0,374.0,132.0,433.0,42.0,74.0,549.0,321.0
14777,2014-2015,Arunachal Pradesh,West Kameng,Nafra,Mathow,85.0,85.0,0.0,167.0,0.0,167.0,77.0,33.0,0.0,125.0,0.0,125.0,46.0
14790,2014-2015,Arunachal Pradesh,West Kameng,Nafra,Mathow,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
16512,2014-2015,Assam,Karbi Anglong,Samelangso,No Panchayat,14760.0,14747.0,610.0,27890.0,1760.0,30260.0,14885.0,10407.0,201.0,11082.0,589.0,11872.0,5394.0
16513,2014-2015,Assam,Karbi Anglong,Samelangso,No Panchayat,6.0,0.0,1.0,63.0,7.0,71.0,37.0,0.0,0.0,0.0,0.0,0.0,0.0
40722,2014-2015,Gujarat,Bharuch,Netrang,Moriyana,210.0,210.0,12.0,285.0,7.0,304.0,147.0,46.0,0.0,39.0,2.0,41.0,19.0
40723,2014-2015,Gujarat,Bharuch,Netrang,Moriyana,584.0,579.0,0.0,617.0,14.0,631.0,322.0,267.0,0.0,44.0,0.0,44.0,14.0
40753,2014-2015,Gujarat,Bharuch,Netrang,Zarna,0.0,0.0,0.0,452.0,9.0,461.0,176.0,0.0,0.0,43.0,0.0,43.0,15.0
40754,2014-2015,Gujarat,Bharuch,Netrang,Zarna,424.0,424.0,2.0,1045.0,18.0,1065.0,517.0,50.0,0.0,10.0,0.0,10.0,5.0


Let's look at a few specific GPs to classify the duplication pattern, just as we did for Dataset 1.

In [28]:
# Pull the first three duplicate GP keys to inspect their patterns
first_dup_keys = (
    Workers_dup[['state_name','district_name','block_name','gp_name','year']]
    .drop_duplicates(subset=['state_name','district_name','block_name','gp_name'])
    .head(3)
)

for _, row in first_dup_keys.iterrows():
    mask = (
        (Workers['state_name']   == row['state_name'])  &
        (Workers['district_name']== row['district_name']) &
        (Workers['block_name']   == row['block_name'])  &
        (Workers['gp_name']      == row['gp_name'])
    )
    display(Workers[mask][['year','state_name','district_name','block_name','gp_name'] + worker_numcols])

,year,state_name,district_name,block_name,gp_name,applied_jobcards,issued_jobcards,registered_workers_scs,registered_workers_sts,registered_workers_others,total_registered_workers,registered_workers_women,active_jobcards,active_workers_scs,active_workers_sts,active_workers_others,total_active_workers,active_workers_women
3132,2014-2015,Andhra Pradesh,Cuddapah,Chitvel,Chitvel,590.0,590.0,6.0,3.0,609.0,618.0,371.0,0.0,1.0,3.0,367.0,371.0,239.0
3133,2014-2015,Andhra Pradesh,Cuddapah,Chitvel,Chitvel,1235.0,1235.0,485.0,45.0,123.0,653.0,374.0,132.0,433.0,42.0,74.0,549.0,321.0
268118,2015-2016,Andhra Pradesh,Cuddapah,Chitvel,Chitvel,615.0,615.0,6.0,3.0,609.0,618.0,371.0,103.0,1.0,3.0,367.0,371.0,239.0
268119,2015-2016,Andhra Pradesh,Cuddapah,Chitvel,Chitvel,655.0,655.0,485.0,45.0,123.0,653.0,374.0,158.0,433.0,42.0,74.0,549.0,321.0


,year,state_name,district_name,block_name,gp_name,applied_jobcards,issued_jobcards,registered_workers_scs,registered_workers_sts,registered_workers_others,total_registered_workers,registered_workers_women,active_jobcards,active_workers_scs,active_workers_sts,active_workers_others,total_active_workers,active_workers_women
14777,2014-2015,Arunachal Pradesh,West Kameng,Nafra,Mathow,85.0,85.0,0.0,167.0,0.0,167.0,77.0,33.0,0.0,125.0,0.0,125.0,46.0
14790,2014-2015,Arunachal Pradesh,West Kameng,Nafra,Mathow,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
279763,2015-2016,Arunachal Pradesh,West Kameng,Nafra,Mathow,85.0,85.0,0.0,167.0,0.0,167.0,77.0,33.0,0.0,125.0,0.0,125.0,46.0
279776,2015-2016,Arunachal Pradesh,West Kameng,Nafra,Mathow,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
546789,2016-2017,Arunachal Pradesh,West Kameng,Nafra,Mathow,83.0,83.0,0.0,167.0,0.0,167.0,77.0,33.0,0.0,125.0,0.0,125.0,46.0
546801,2016-2017,Arunachal Pradesh,West Kameng,Nafra,Mathow,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
813659,2017-2018,Arunachal Pradesh,West Kameng,Nafra,Mathow,74.0,74.0,0.0,167.0,0.0,167.0,77.0,34.0,0.0,125.0,0.0,125.0,46.0
813671,2017-2018,Arunachal Pradesh,West Kameng,Nafra,Mathow,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1080839,2018-2019,Arunachal Pradesh,West Kameng,Nafra,Mathow,48.0,48.0,0.0,167.0,0.0,167.0,77.0,33.0,0.0,125.0,0.0,125.0,46.0
1080851,2018-2019,Arunachal Pradesh,West Kameng,Nafra,Mathow,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0


,year,state_name,district_name,block_name,gp_name,applied_jobcards,issued_jobcards,registered_workers_scs,registered_workers_sts,registered_workers_others,total_registered_workers,registered_workers_women,active_jobcards,active_workers_scs,active_workers_sts,active_workers_others,total_active_workers,active_workers_women
16512,2014-2015,Assam,Karbi Anglong,Samelangso,No Panchayat,14760.0,14747.0,610.0,27890.0,1760.0,30260.0,14885.0,10407.0,201.0,11082.0,589.0,11872.0,5394.0
16513,2014-2015,Assam,Karbi Anglong,Samelangso,No Panchayat,6.0,0.0,1.0,63.0,7.0,71.0,37.0,0.0,0.0,0.0,0.0,0.0,0.0
281498,2015-2016,Assam,Karbi Anglong,Samelangso,No Panchayat,15829.0,15692.0,610.0,27890.0,1760.0,30260.0,14885.0,10617.0,201.0,11082.0,589.0,11872.0,5394.0
281499,2015-2016,Assam,Karbi Anglong,Samelangso,No Panchayat,26.0,0.0,1.0,63.0,7.0,71.0,37.0,0.0,0.0,0.0,0.0,0.0,0.0
548673,2016-2017,Assam,Karbi Anglong,Samelangso,No Panchayat,15227.0,15103.0,610.0,27890.0,1760.0,30260.0,14885.0,10995.0,201.0,11082.0,589.0,11872.0,5394.0
548674,2016-2017,Assam,Karbi Anglong,Samelangso,No Panchayat,29.0,0.0,1.0,63.0,7.0,71.0,37.0,0.0,0.0,0.0,0.0,0.0,0.0
815523,2017-2018,Assam,Karbi Anglong,Samelangso,No Panchayat,12750.0,12403.0,610.0,27890.0,1760.0,30260.0,14885.0,11555.0,201.0,11082.0,589.0,11872.0,5394.0
815524,2017-2018,Assam,Karbi Anglong,Samelangso,No Panchayat,29.0,0.0,1.0,63.0,7.0,71.0,37.0,0.0,0.0,0.0,0.0,0.0,0.0
1082703,2018-2019,Assam,Karbi Anglong,Samelangso,No Panchayat,13305.0,12918.0,610.0,27890.0,1760.0,30260.0,14885.0,11915.0,201.0,11082.0,589.0,11872.0,5394.0
1082704,2018-2019,Assam,Karbi Anglong,Samelangso,No Panchayat,29.0,0.0,1.0,63.0,7.0,71.0,37.0,0.0,0.0,0.0,0.0,0.0,0.0


These results reveal the same three duplication patterns as Dataset 1.

**Case 1 (Mathow, Arunachal Pradesh):** A real row paired with an all-zero ghost row, consistently across every year. The zero row likely represents a residual administrative entry with no actual data. The zero-row filter removes it cleanly.

**Case 2 (No Panchayat, Assam – Karbi Anglong):** One large dominant row paired with a small but non-zero satellite row (e.g. 6 applied job cards vs 14,760). The satellite is not all-zero so the filter alone does not drop it. However, since the dominant row is strictly larger on every column, `max()` effectively preserves the main row in full — the satellite contributes nothing.

**Case 3 (Chitvel, Andhra Pradesh):** Two rows with substantially different and plausible caste breakdowns — neither is a zero ghost nor a clear satellite. This is genuine split reporting: two administrative sub-units recorded under the same GP name. Taking `max()` column-wise mixes values from different rows (e.g. it would pair the larger SC count from one row with the larger Others count from another), which risks creating internally inconsistent caste totals. This is a known limitation of the `max()` approach and should be kept in mind when interpreting caste-disaggregated outcomes for affected GPs.

Applying the same two-step resolution used for Dataset 1: first drop all-zero rows (ghost entries), then take `max()` within each GP-year group for any remaining duplicates. Since the worker counts accumulate over a reporting period, the maximum is the best approximation of the true final value.

In [29]:
Workers_clean = Workers.copy()

# Step 1: drop rows where all numeric columns are zero
Workers_clean = Workers_clean[Workers_clean[worker_numcols].sum(axis=1) > 0]
print(f"Rows after dropping all-zero rows: {len(Workers_clean)}")

# Step 2: collapse remaining duplicates by taking max within each GP-year
gp_key = ['year','state_name','district_name','block_name','gp_name']
Workers_clean = Workers_clean.groupby(gp_key, as_index=False)[worker_numcols].max()
print(f"Rows after deduplication: {len(Workers_clean)}")

Rows after dropping all-zero rows: 2651938
Rows after deduplication: 2651519


In [30]:
# Confirm all keys are now unique
Workers_clean[['year','state_name','district_name','block_name','gp_name']].value_counts().unique()

array([1])

Now all keys are unique!

### 2.4 Selecting necessary columns

In [31]:
WorkersClean = Workers_clean[gp_key + worker_numcols].drop(columns = ['registered_workers_women', 'active_workers_women'])
WorkersClean.head()

,year,state_name,district_name,block_name,gp_name,applied_jobcards,issued_jobcards,registered_workers_scs,registered_workers_sts,registered_workers_others,total_registered_workers,active_jobcards,active_workers_scs,active_workers_sts,active_workers_others,total_active_workers
0,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Campbell Bay,376.0,375.0,0.0,0.0,718.0,718.0,137.0,0.0,0.0,10.0,10.0
1,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Govindnagar,429.0,429.0,0.0,0.0,790.0,790.0,258.0,0.0,0.0,51.0,51.0
2,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Great & Little Nicobar,0.0,0.0,0.0,433.0,1.0,434.0,0.0,0.0,0.0,0.0,0.0
3,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Laxmi Nagar,460.0,460.0,0.0,0.0,922.0,922.0,237.0,0.0,0.0,24.0,24.0
4,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Chowra Tc,309.0,309.0,0.0,646.0,2.0,648.0,150.0,0.0,63.0,2.0,65.0


registered_workers_scs - Number of registered workers from SCs <br>
registered_workers_sts - Number of registered workers from STs <br>
registered_workers_others - Number of registered workers from other categories <br>
total_registered_workers - Total number of registered workers <br>

active_workers_scs - Number of active workers from SCs <br>
active_workers_sts - Number of active workers from STs <br>
active_workers_others - Number of active workers from other categories <br>
total_active_workers - Total number of active workers

### 2.5 Internal consistency: do SC + ST + Others add up to Total?

Since `total_registered_workers` and `total_active_workers` is reported separately from the three caste sub-categories, we can verify whether `sc + sts + others == tot`. Large discrepancies would directly undermine any equity analysis built on these variables.

In [32]:
WorkersClean['registered_comp_sum'] = (
    WorkersClean['registered_workers_scs'] +
    WorkersClean['registered_workers_sts'] +
    WorkersClean['registered_workers_others']
)
WorkersClean['reg_discrepancy'] = WorkersClean['total_registered_workers'] - WorkersClean['registered_comp_sum']

WorkersClean['active_comp_sum'] = (
    WorkersClean['active_workers_scs'] +
    WorkersClean['active_workers_sts'] +
    WorkersClean['active_workers_others']
)
WorkersClean['active_discrepancy'] = WorkersClean['total_active_workers'] - WorkersClean['active_comp_sum']

reg_inconsistent = (WorkersClean['reg_discrepancy'].abs() > 0).sum()
print(f"Rows where registered sc + sts + others != tot: {reg_inconsistent} ({100*reg_inconsistent/len(WorkersClean):.2f}%)")
print()
print("Distribution of discrepancy (tot - sum of components):")
print(WorkersClean['reg_discrepancy'].describe())

print('---------------------------------------------------------------------------')
active_inconsistent = (WorkersClean['active_discrepancy'].abs() > 0).sum()
print(f"Rows where active sc + sts + others != tot: {active_inconsistent} ({100*active_inconsistent/len(WorkersClean):.2f}%)")
print()
print("Distribution of discrepancy (tot - sum of components):")
print(WorkersClean['active_discrepancy'].describe())


Rows where registered sc + sts + others != tot: 89 (0.00%)

Distribution of discrepancy (tot - sum of components):
count    2.651519e+06
mean    -3.722395e-03
std      1.073558e+00
min     -4.860000e+02
25%      0.000000e+00
50%      0.000000e+00
75%      0.000000e+00
max      0.000000e+00
Name: reg_discrepancy, dtype: float64
---------------------------------------------------------------------------
Rows where active sc + sts + others != tot: 58 (0.00%)

Distribution of discrepancy (tot - sum of components):
count    2.651519e+06
mean    -1.436535e-03
std      5.802460e-01
min     -2.930000e+02
25%      0.000000e+00
50%      0.000000e+00
75%      0.000000e+00
max      0.000000e+00
Name: active_discrepancy, dtype: float64


Mostly the discrepancies are small (median is 0, mean is around 1-2), but there are some large outliers (max discrepancy is around 300-400 in both cases) which may warrant further investigation.

Let's sort out the rows where the discrepancies are greater than 5% of the total number reported

In [33]:
WorkersClean['reg_gap_pct'] = (
    (WorkersClean['reg_discrepancy'] / WorkersClean['total_registered_workers'])
)

WorkersClean['active_gap_pct'] = (
    (WorkersClean['active_discrepancy'] / WorkersClean['total_active_workers'])
)

reg_inconsistent = (WorkersClean['reg_gap_pct'].abs() > 0.05).sum()
print(f"Rows where reg inconsistency is greater than 5% : {reg_inconsistent} ({100*reg_inconsistent/len(WorkersClean):.2f}%)")
print()
active_inconsistent = (WorkersClean['active_gap_pct'].abs() > 0.05).sum()
print(f"Rows where active inconsistency is greater than 5% : {active_inconsistent} ({100*active_inconsistent/len(WorkersClean):.2f}%)")
print()


Rows where reg inconsistency is greater than 5% : 73 (0.00%)

Rows where active inconsistency is greater than 5% : 46 (0.00%)



A very small % of PCs display discrepancies greater than 5%. Therefore, we can just drop these rows for our analysis.

In [34]:
Workers_filtered = WorkersClean[(WorkersClean['reg_gap_pct'].abs() <= 0.05) & 
                             (WorkersClean['active_gap_pct'].abs() <= 0.05)].drop(
                                 columns = ['registered_comp_sum', 'reg_discrepancy', 'reg_gap_pct',
                                            'active_comp_sum', 'active_discrepancy', 'active_gap_pct']
                             )
Workers_filtered.head()

,year,state_name,district_name,block_name,gp_name,applied_jobcards,issued_jobcards,registered_workers_scs,registered_workers_sts,registered_workers_others,total_registered_workers,active_jobcards,active_workers_scs,active_workers_sts,active_workers_others,total_active_workers
0,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Campbell Bay,376.0,375.0,0.0,0.0,718.0,718.0,137.0,0.0,0.0,10.0,10.0
1,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Govindnagar,429.0,429.0,0.0,0.0,790.0,790.0,258.0,0.0,0.0,51.0,51.0
3,2014-2015,Andaman And Nicobar,Nicobars,Campbell Bay,Laxmi Nagar,460.0,460.0,0.0,0.0,922.0,922.0,237.0,0.0,0.0,24.0,24.0
4,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Chowra Tc,309.0,309.0,0.0,646.0,2.0,648.0,150.0,0.0,63.0,2.0,65.0
5,2014-2015,Andaman And Nicobar,Nicobars,Nancowry,Kamorta Tc,612.0,612.0,0.0,1255.0,191.0,1446.0,388.0,0.0,443.0,15.0,458.0


### 2.6 Missing values

Checking for NaN values across all columns in `Workers_filtered`. Even though the zero-row filter in 2.3 removed all-zero rows, it is possible that individual cells are NaN rather than zero — particularly in the caste-disaggregated columns, which are the key equity variables.

In [35]:
missing = Workers_filtered.isnull().sum()
missing_pct = 100 * missing / len(Workers_filtered)

missing_summary = pd.DataFrame({'missing_count': missing, 'missing_pct': missing_pct.round(2)})
display(missing_summary[missing_summary['missing_count'] > 0])

print(f"\nTotal rows in Workers_filtered: {len(Workers_filtered)}")
print(f"Rows with any missing value: {Workers_filtered.isnull().any(axis=1).sum()}")


,missing_count,missing_pct



Total rows in Workers_filtered: 2460169
Rows with any missing value: 0


### 2.7 State × year coverage: GP counts

A pivot table showing the number of GPs observed per state per year. This reveals whether coverage is stable across the panel — sudden drops or gaps in a state-year cell indicate that data was not reported or not collected for that year, which matters for any panel fixed-effects analysis.

In [36]:
coverage = Workers_filtered.pivot_table(
    index='state_name',
    columns='year',
    values='gp_name',
    aggfunc='count'
)

# Sort years chronologically (they are strings like "2014-2015")
coverage = coverage[sorted(coverage.columns)]

# Highlight missing (NaN) cells and large drops to make gaps visible
coverage.style.background_gradient(cmap='YlGn', axis=1).format('{:.0f}', na_rep='—')


year,2014-2015,2015-2016,2016-2017,2017-2018,2018-2019,2019-2020,2020-2021,2021-2022,2022-2023,2023-2024
state_name,,,,,,,,,,
Andaman And Nicobar,74,74,72,72,72,72,72,72,72,72
Andhra Pradesh,12831,12831,13249,13246,13246,11808,13242,13242,13242,13241
Arunachal Pradesh,1784,1784,1903,1903,1903,1903,1903,1903,1903,1903
Assam,2619,2619,2619,2619,2619,2619,2619,2619,2619,2619
Bihar,8358,8358,8323,8323,8324,8323,8323,8323,8322,8321
Chhattisgarh,9716,9716,9781,9781,9780,9780,9779,9779,9779,9779
Dn Haveli And Dd,1,1,1,1,1,1,1,1,1,1
Goa,137,137,137,137,137,137,137,137,137,137
Gujarat,13824,13821,13635,13635,13635,13637,13637,13637,13637,13456


Let's take a look at how the data cleaning reshaped our data:

In [37]:
print(f"\nTotal rows and columns in Raw Data (Workers): {Workers.shape}")
print(f"\nTotal rows and columns in Cleaned Data (Workers_filtered): {Workers_filtered.shape}")

print(f'{len(Workers) - len(Workers_filtered)} rows ({100*(len(Workers) - len(Workers_filtered))/len(Workers):.2f}%) removed during cleaning')


Total rows and columns in Raw Data (Workers): (2676899, 20)

Total rows and columns in Cleaned Data (Workers_filtered): (2460169, 16)
216730 rows (8.10%) removed during cleaning


In [62]:
Workers_filtered.to_csv('/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/3. Processed/Workers_filtered.csv', index=False)

## Dataset 3: Mittal - SC Inequity, ST Inequity, mapping

First we load Mittal's SC and ST inequity datasets, along with the panchayat to MIS index mapping dataset.

In [38]:
SC_inequity = pd.read_csv("/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/1. Raw/Mittal/panchayat_level_sc_inequity.csv")
print("SC Inequity:")
display(SC_inequity.head())

ST_inequity = pd.read_csv("/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/1. Raw/Mittal/panchayat_level_st_inequity.csv")
print("ST Inequity:")
display(ST_inequity.head())

SC Inequity:


,Unnamed: 0,mis_index,equity_score1,equity_score2,equity_score3,equity_score4,category,equity_score5,inequity_score
0,0,166463,0.018656,0.004735,0.094903,0.189998,3,0.084442,1.763303
1,1,166475,0.597425,-0.161602,0.302553,0.005162,2,0.015320,1.619678
2,2,166478,0.103851,NaN,0.120494,0.120689,3,0.022218,1.982577
3,3,166479,0.469600,-0.351881,0.468233,0.243940,3,0.064675,1.531420
4,4,166533,0.278150,0.551750,0.388139,0.166026,2,0.053694,1.289105


ST Inequity:


,Unnamed: 0,mis_index,equity_score1,equity_score2,equity_score3,equity_score4,category,equity_score5,inequity_score
0,0,166576,-0.003630,NaN,-0.059030,0.417278,5,0.240800,1.793128
1,1,166739,0.486835,NaN,0.363058,-0.288619,5,-0.021083,1.779054
2,2,166743,0.536776,0.502125,0.595711,0.000000,5,0.342944,1.047495
3,3,166920,0.280412,NaN,0.295163,0.000000,6,-0.033988,1.838920
4,4,166933,-0.044191,0.519517,-0.035219,0.000000,6,0.020523,1.767770


In [39]:
print("SC_inequity")
print(SC_inequity.info())

SC_inequity
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45218 entries, 0 to 45217
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      45218 non-null  int64  
 1   mis_index       45218 non-null  int64  
 2   equity_score1   25086 non-null  float64
 3   equity_score2   21582 non-null  float64
 4   equity_score3   25086 non-null  float64
 5   equity_score4   45218 non-null  float64
 6   category        45218 non-null  int64  
 7   equity_score5   45218 non-null  float64
 8   inequity_score  45218 non-null  float64
dtypes: float64(6), int64(3)
memory usage: 3.1 MB
None


In [40]:
print("ST_inequity")
print(ST_inequity.info())

ST_inequity
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 26426 entries, 0 to 26425
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      26426 non-null  int64  
 1   mis_index       26426 non-null  int64  
 2   equity_score1   10107 non-null  float64
 3   equity_score2   11446 non-null  float64
 4   equity_score3   10107 non-null  float64
 5   equity_score4   26426 non-null  float64
 6   category        26426 non-null  int64  
 7   equity_score5   26426 non-null  float64
 8   inequity_score  26426 non-null  float64
dtypes: float64(6), int64(3)
memory usage: 1.8 MB
None


### 3.1 Combining SC and ST Inequity

The SC and ST panchayat-level datasets are mutually exclusive — no panchayat appears in both. They are stacked into a single dataframe with a `group` column indicating the source dataset (SC or ST). The existing `inequity_score` column serves as the overall inequity metric for each panchayat.

In [41]:
SC_inequity["group"] = "SC"
ST_inequity["group"] = "ST"

panchayat_inequity = pd.concat([SC_inequity, ST_inequity], ignore_index=True)

display(panchayat_inequity.head(10))
print(f"Total panchayats: {len(panchayat_inequity)}")
print(f"SC panchayats: {(panchayat_inequity['group'] == 'SC').sum()}")
print(f"ST panchayats: {(panchayat_inequity['group'] == 'ST').sum()}")
panchayat_inequity['category'].value_counts()

,Unnamed: 0,mis_index,equity_score1,equity_score2,equity_score3,equity_score4,category,equity_score5,inequity_score,group
0,0,166463,0.018656,0.004735,0.094903,0.189998,3,0.084442,1.763303,SC
1,1,166475,0.597425,-0.161602,0.302553,0.005162,2,0.015320,1.619678,SC
2,2,166478,0.103851,NaN,0.120494,0.120689,3,0.022218,1.982577,SC
3,3,166479,0.469600,-0.351881,0.468233,0.243940,3,0.064675,1.531420,SC
4,4,166533,0.278150,0.551750,0.388139,0.166026,2,0.053694,1.289105,SC
5,5,166597,0.756245,0.230070,0.591301,0.075920,3,0.012184,1.175621,SC
6,6,166606,0.482028,NaN,0.352771,0.068385,3,0.040465,1.564980,SC
7,7,166608,0.016419,NaN,0.061557,0.189844,3,0.043494,2.021064,SC
8,8,166626,0.011295,NaN,0.191799,0.444023,2,0.099024,1.717966,SC
9,9,166653,0.532565,NaN,0.518486,-0.282379,3,-0.082128,1.639313,SC


Total panchayats: 71644
SC panchayats: 45218
ST panchayats: 26426


category
3    17272
2    16867
1    11079
5     9933
4     8258
6     8235
Name: count, dtype: int64

Category explanations:<br>

1 – **Consolidated significant SC panchayats:** PCs where SC households are geographically concentrated together (high SC density, clustered) <br>
2 – **Segregated significant SC panchayats:** PCs with significant SC population but where SC households are spatially separated/segregated from general caste households <br>
3 – **Mixed SC panchayats:** Panchayats with significant SC population that is neither consolidated nor segregated — more integrated settlement patterns <br>

4 – **Consolidated significant ST panchayats:** Same as Category 1 but for Scheduled Tribe households <br>
5 – **Segregated significant ST panchayats:** Same as Category 2 but for Scheduled Tribe households <br>
6 – **Mixed ST panchayats:** Same as Category 3 but for Scheduled Tribe households

In [63]:
panchayat_inequity.to_csv('/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/2. Interim/Panchayat_inequity.csv', index=False)

### 3.2 Mapping Dataset

In [42]:
PC_MIS_mapping = pd.read_csv('/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/2. Interim/Mittal/V1_panchayat_ID_MIS_index_mapping.csv')
PC_MIS_mapping.head()

,Panchayat_ID,Gram_panchayat_name,Subdistrict census code,Subdistrict name,District census code,District name,State census code,State name,mis_index
0,221810,pitchikalapadu,4730,Yetapaka,541,visakhapatnam,28,andhra pradesh,NaN
1,221811,kannaigudem,4730,Yetapaka,541,visakhapatnam,28,andhra pradesh,NaN
2,221812,laxmipuram,4730,Yetapaka,541,visakhapatnam,28,andhra pradesh,NaN
3,221813,ramagopalapuram,4730,Yetapaka,541,visakhapatnam,28,andhra pradesh,NaN
4,221814,nallakunta,4730,Yetapaka,541,visakhapatnam,28,andhra pradesh,NaN


In [43]:
PC_MIS_mapping.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 218350 entries, 0 to 218349
Data columns (total 9 columns):
 #   Column                   Non-Null Count   Dtype 
---  ------                   --------------   ----- 
 0   Panchayat_ID             217930 non-null  object
 1   Gram_panchayat_name      217458 non-null  object
 2   Subdistrict census code  217930 non-null  object
 3   Subdistrict name         217930 non-null  object
 4   District census code     217930 non-null  object
 5   District name            217619 non-null  object
 6   State census code        217930 non-null  object
 7   State name               217930 non-null  object
 8   mis_index                115223 non-null  object
dtypes: object(9)
memory usage: 15.0+ MB


### 3.3 Data Quality Issue: Corrupted Rows in Mapping File

`PC_MIS_mapping.info()` shows all columns read as `object` type, which is unexpected for the ID and census code columns. Inspection reveals **420 rows where every column contains a GPS coordinate pair** (e.g. `87.20157 25.82769`) instead of the expected census identifiers. These rows appear to be GPS traces that were incorrectly appended to the CSV during file creation — every field in such a row is a `longitude latitude` string rather than a panchayat ID or name.

**Fix:** Drop all rows where `Panchayat_ID` cannot be parsed as a number, then cast `Panchayat_ID` and the three census code columns to `Int64`.

In [44]:
# Drop corrupted GPS-coordinate rows
PC_MIS_mapping = PC_MIS_mapping[
    pd.to_numeric(PC_MIS_mapping['Panchayat_ID'], errors='coerce').notna()
].copy()

# Cast ID and census code columns to integer
int_cols = ['Panchayat_ID', 'Subdistrict census code', 'District census code', 'State census code']
PC_MIS_mapping[int_cols] = PC_MIS_mapping[int_cols].apply(pd.to_numeric).astype('Int64')

print(f'Rows after dropping corrupted entries: {len(PC_MIS_mapping):,}')
PC_MIS_mapping.info()

Rows after dropping corrupted entries: 217,510
<class 'pandas.core.frame.DataFrame'>
Index: 217510 entries, 0 to 218349
Data columns (total 9 columns):
 #   Column                   Non-Null Count   Dtype 
---  ------                   --------------   ----- 
 0   Panchayat_ID             217510 non-null  Int64 
 1   Gram_panchayat_name      217038 non-null  object
 2   Subdistrict census code  217510 non-null  Int64 
 3   Subdistrict name         217510 non-null  object
 4   District census code     217510 non-null  Int64 
 5   District name            217199 non-null  object
 6   State census code        217510 non-null  Int64 
 7   State name               217510 non-null  object
 8   mis_index                114806 non-null  object
dtypes: Int64(4), object(5)
memory usage: 17.4+ MB


#### Issue 1: Exact Duplicate Rows

In [45]:
#dplicated PC ids
dup_ids = PC_MIS_mapping[PC_MIS_mapping.duplicated(subset=['Panchayat_ID'], keep=False)]

print(f"Duplicates: {dup_ids.shape[0]}")          # total duplicate rows (should be ~16356)
print(f"Unique IDs: {dup_ids['Panchayat_ID'].nunique()}")   # unique IDs (~8178)

Duplicates: 16356
Unique IDs: 8178


Seems like the duplicates might be perfect copies. Let's verify this.

In [46]:
dup_ids.sort_values('Panchayat_ID').head(10)

,Panchayat_ID,Gram_panchayat_name,Subdistrict census code,Subdistrict name,District census code,District name,State census code,State name,mis_index
17835,99300,balmiki nagar,1013,Sidhaw,203,pashchim champaran,10,bihar,NaN
26019,99300,balmiki nagar,1013,Sidhaw,203,pashchim champaran,10,bihar,NaN
17836,99301,laxmipur rampurwa,1013,Sidhaw,203,pashchim champaran,10,bihar,NaN
26020,99301,laxmipur rampurwa,1013,Sidhaw,203,pashchim champaran,10,bihar,NaN
26021,99302,santpur sohriya,1013,Sidhaw,203,pashchim champaran,10,bihar,NaN
17837,99302,santpur sohriya,1013,Sidhaw,203,pashchim champaran,10,bihar,NaN
26022,99303,champapur gonauli,1013,Sidhaw,203,pashchim champaran,10,bihar,NaN
17838,99303,champapur gonauli,1013,Sidhaw,203,pashchim champaran,10,bihar,NaN
17839,99304,mahuawa katharwa,1013,Sidhaw,203,pashchim champaran,10,bihar,NaN
26023,99304,mahuawa katharwa,1013,Sidhaw,203,pashchim champaran,10,bihar,NaN


16,356 rows share a duplicated `Panchayat_ID` (8,178 affected IDs). Every duplicate pair is **identical across all columns**, including `mis_index` — these are not conflicting records but verbatim copies, likely from double-appending a data block during file creation. They are safe to drop with `drop_duplicates()`.

In [47]:
PC_MIS_mapping = PC_MIS_mapping.drop_duplicates(subset='Panchayat_ID').copy()
print(f'Rows after dropping exact duplicates: {len(PC_MIS_mapping):,}')
print(f'Unique Panchayat_IDs: {PC_MIS_mapping["Panchayat_ID"].nunique():,}')

Rows after dropping exact duplicates: 209,332
Unique Panchayat_IDs: 209,332


#### Issue 2: `mis_index` Still Stored as Object

After the GPS-row removal, `mis_index` remains `object` type because the column contained mixed strings and numbers before cleaning. It should be cast to `float`. The ~47% missingness is expected — only 115k of 217k panchayats have a non-null MIS index (documented in CLAUDE.md).

In [48]:
PC_MIS_mapping['mis_index'] = pd.to_numeric(PC_MIS_mapping['mis_index'], errors='coerce')
print(f'mis_index dtype: {PC_MIS_mapping["mis_index"].dtype}')
print(f'Missing mis_index: {PC_MIS_mapping["mis_index"].isna().sum():,} ({100*PC_MIS_mapping["mis_index"].isna().mean():.1f}%)')

mis_index dtype: float64
Missing mis_index: 100,043 (47.8%)


#### Issue 3: Inconsistent String Formatting

`Gram_panchayat_name` has 101 entries with leading or trailing whitespace, and `Subdistrict name` is in mixed/title case while `District name` and `State name` are already lowercase. Stripping whitespace and lowercasing all name columns ensures consistent matching when this file is joined to other datasets.

In [49]:
str_cols = ['Gram_panchayat_name', 'Subdistrict name', 'District name', 'State name']
PC_MIS_mapping[str_cols] = (
    PC_MIS_mapping[str_cols].apply(lambda s: s.str.strip().str.lower())
)
PC_MIS_mapping.info()

<class 'pandas.core.frame.DataFrame'>
Index: 209332 entries, 0 to 218349
Data columns (total 9 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Panchayat_ID             209332 non-null  Int64  
 1   Gram_panchayat_name      208901 non-null  object 
 2   Subdistrict census code  209332 non-null  Int64  
 3   Subdistrict name         209332 non-null  object 
 4   District census code     209332 non-null  Int64  
 5   District name            209021 non-null  object 
 6   State census code        209332 non-null  Int64  
 7   State name               209332 non-null  object 
 8   mis_index                109289 non-null  float64
dtypes: Int64(4), float64(1), object(4)
memory usage: 16.8+ MB


In [64]:
PC_MIS_mapping.to_csv('/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/2. Interim/Cleaned_PC_MIS_mapping.csv', index=False)

## Dataset 4: SHRUG 2001 and 2011

In [50]:
pop_census_2001 = pd.read_csv('/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/1. Raw/SHRUG/2001 population census rural/pc01_vd_clean_pc01subdist.csv')
print("Population Census 2001:")
display(pop_census_2001.head())

Population Census 2001:


,pc01_state_id,pc01_district_id,pc01_subdistrict_id,pc01_vd_area,pc01_vd_t_hh,pc01_vd_t_p,pc01_vd_t_m,pc01_vd_t_f,pc01_vd_sc_p,pc01_vd_sc_m,...,pc01_vd_tw_w_el,pc01_vd_tank_irr,pc01_vd_river_irr,pc01_vd_lake_irr,pc01_vd_w_fall,pc01_vd_oth_irr,pc01_vd_tot_irr,pc01_vd_un_irr,pc01_vd_cult_waste,pc01_vd_area_na_cu
0,1,1,1,29784.0,48862,380302,198149,182153,26,26,...,0.0,0.0,0.0,0.0,0.0,139.630000,12434.019948,7369.540019,4756.200012,5030.120025
1,1,1,2,28345.5,28816,201825,104250,97575,8,8,...,0.0,0.0,0.0,0.0,0.0,0.000000,8866.730016,13119.170033,3824.990018,2215.120004
2,1,1,3,6988.5,6192,42771,23797,18974,27,27,...,0.0,0.0,0.0,0.0,0.0,519.229999,876.170001,1496.090001,1736.600005,2796.579984
3,1,2,1,4466.2,3843,30144,17617,12527,103,94,...,0.0,0.0,0.0,0.0,0.0,0.000000,596.900005,619.700002,2279.900013,942.599995
4,1,2,2,10280.4,16102,109576,57413,52163,15,15,...,0.0,0.0,0.0,0.0,0.0,0.000000,4147.600015,2470.499988,1123.700015,2511.100006


### 4.1 Loading all necessary mapping datasets

Next let's load the SHRUG keys dataset which helps map the given datasets' keys to census identifier numbers. We'll have to first use `shrid2` to get the state, district, and subdistrict names.

In [51]:
shrid_loc =pd.read_csv('/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/1. Raw/SHRUG/shrug-shrid-keys-csv/shrid_loc_names.csv')
print("Location Names:")
display(shrid_loc.head())

Location Names:


,shrid2,state_name,district_name,subdistrict_name,town_name,village_name,place_name
0,11-01-001-00001-000001,jammu kashmir,kupwara,kupwara,NaN,bore,bore
1,11-01-001-00001-000002,jammu kashmir,kupwara,kupwara,NaN,keran,keran
2,11-01-001-00001-000005,jammu kashmir,kupwara,kupwara,NaN,mindiyan,mindiyan
3,11-01-001-00001-000006,jammu kashmir,kupwara,kupwara,NaN,patrin,patrin
4,11-01-001-00001-000007,jammu kashmir,kupwara,kupwara,NaN,juma gund,juma gund


In [52]:
shrid_subdistrict = pd.read_csv("/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/1. Raw/SHRUG/shrug-pc-keys-csv/shrid_pc01subdist_key.csv")
shrid_subdistrict.head()

,shrid2,pc01_state_id,pc01_district_id,pc01_subdistrict_id
0,01-03-01-0002-00008200,3,1,2
1,01-03-01-0002-00008300,3,1,2
2,01-03-01-0002-00009300,3,1,2
3,01-03-01-0002-00009500,3,1,2
4,01-03-01-0002-00009700,3,1,2


### 4.2 Mapping census data to state, district, and subdistrict names

**Approach:** `shrid_subdistrict` maps each village (`shrid2`) to internal SHRUG numeric IDs (`pc01_state_id`, `pc01_district_id`, `pc01_subdistrict_id`). Note these are **not** the same as the census codes in `PC_MIS_mapping` — shrid uses district-relative subdistrict numbering (1, 2, 3 per district) while PC_MIS uses state-relative codes (e.g. 200, 201...). The common key for the eventual merge into `panchayat_inequity` will therefore be **string names** (state, district, subdistrict), not numeric IDs.

`shrid_loc` maps the same `shrid2` to human-readable names. Joining these two on `shrid2` and collapsing to the subdistrict level gives a name-keyed lookup. That lookup is then merged with `pop_census_2001` (keyed by the same SHRUG numeric IDs) to attach population data to named subdistricts.

**Name conflicts:** 919 subdistricts have more than one `subdistrict_name` across their villages in `shrid_loc` (minor spelling variants). The most frequent name per subdistrict is taken as canonical. 6,701 villages in `shrid_subdistrict` have no match in `shrid_loc`; these contribute NaN names and are handled with a fallback in the aggregation.

In [53]:
subdist_key = ['pc01_state_id', 'pc01_district_id', 'pc01_subdistrict_id']

# Step 1: join village-level shrids to their names
shrid_merged = shrid_subdistrict.merge(
    shrid_loc[['shrid2', 'state_name', 'district_name', 'subdistrict_name']],
    on='shrid2', how='left'
)

# Step 2: collapse to subdistrict level
# mode() raises IndexError when all values in a group are NaN, so we use a safe wrapper
def safe_mode(x):
    m = x.dropna().mode()
    return m.iloc[0] if len(m) > 0 else None

subdist_names = (
    shrid_merged
    .groupby(subdist_key)[['state_name', 'district_name', 'subdistrict_name']]
    .agg(safe_mode)
    .reset_index()
)

print(f'Unique subdistricts in name lookup: {len(subdist_names):,}')
subdist_names.head()

Unique subdistricts in name lookup: 5,483


,pc01_state_id,pc01_district_id,pc01_subdistrict_id,state_name,district_name,subdistrict_name
0,1,1,1,jammu kashmir,kupwara,kupwara
1,1,1,2,jammu kashmir,kupwara,handwara
2,1,1,3,jammu kashmir,kupwara,karnah
3,1,2,1,jammu kashmir,bandipore,gurez
4,1,2,2,jammu kashmir,bandipore,bandipora


Merge the subdistrict name lookup with `pop_census_2001` on the three SHRUG numeric ID columns. All 5,333 subdistricts in `pop_census_2001` are present in `shrid_subdistrict`, so we expect a complete match. The result `subdist_census_2001` is keyed by names and ready to merge into `panchayat_inequity` via `PC_MIS_mapping` on `(state_name, district_name, subdistrict_name)`.

In [54]:
# Step 3: attach census data to named subdistricts
subdist_census_2001 = subdist_names.merge(pop_census_2001, on=subdist_key, how='right')

unmatched = subdist_census_2001['subdistrict_name'].isna().sum()
print(f'Rows in subdist_census_2001: {len(subdist_census_2001):,}')
print(f'Unmatched (no name found): {unmatched}')
subdist_census_2001.head()

Rows in subdist_census_2001: 5,333
Unmatched (no name found): 4


,pc01_state_id,pc01_district_id,pc01_subdistrict_id,state_name,district_name,subdistrict_name,pc01_vd_area,pc01_vd_t_hh,pc01_vd_t_p,pc01_vd_t_m,...,pc01_vd_tw_w_el,pc01_vd_tank_irr,pc01_vd_river_irr,pc01_vd_lake_irr,pc01_vd_w_fall,pc01_vd_oth_irr,pc01_vd_tot_irr,pc01_vd_un_irr,pc01_vd_cult_waste,pc01_vd_area_na_cu
0,1,1,1,jammu kashmir,kupwara,kupwara,29784.0,48862,380302,198149,...,0.0,0.0,0.0,0.0,0.0,139.630000,12434.019948,7369.540019,4756.200012,5030.120025
1,1,1,2,jammu kashmir,kupwara,handwara,28345.5,28816,201825,104250,...,0.0,0.0,0.0,0.0,0.0,0.000000,8866.730016,13119.170033,3824.990018,2215.120004
2,1,1,3,jammu kashmir,kupwara,karnah,6988.5,6192,42771,23797,...,0.0,0.0,0.0,0.0,0.0,519.229999,876.170001,1496.090001,1736.600005,2796.579984
3,1,2,1,jammu kashmir,bandipore,gurez,4466.2,3843,30144,17617,...,0.0,0.0,0.0,0.0,0.0,0.000000,596.900005,619.700002,2279.900013,942.599995
4,1,2,2,jammu kashmir,bandipore,bandipora,10280.4,16102,109576,57413,...,0.0,0.0,0.0,0.0,0.0,0.000000,4147.600015,2470.499988,1123.700015,2511.100006


In [65]:
subdist_census_2001.to_csv('/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/2. Interim/Subdistrict_census_2001.csv', index=False)

## 5. Merging Census Data with Inequity index on subdistricts

First let's select only relevant columns from the census data.

In [55]:
subdist_census_2001_filtered = (
    subdist_census_2001[
        ['state_name', 'district_name', 'subdistrict_name',
         'pc01_vd_t_p', 'pc01_vd_sc_p', 'pc01_vd_st_p']
    ]
    .rename(columns={
        'pc01_vd_t_p': 'tot_pop_persons',
        'pc01_vd_sc_p': 'SC_pop_persons',
        'pc01_vd_st_p': 'ST_pop_persons',
    })
)

print(f'Rows: {len(subdist_census_2001_filtered):,}')
subdist_census_2001_filtered.head()

Rows: 5,333


,state_name,district_name,subdistrict_name,tot_pop_persons,SC_pop_persons,ST_pop_persons
0,jammu kashmir,kupwara,kupwara,380302,26,33108
1,jammu kashmir,kupwara,handwara,201825,8,10988
2,jammu kashmir,kupwara,karnah,42771,27,6917
3,jammu kashmir,bandipore,gurez,30144,103,26239
4,jammu kashmir,bandipore,bandipora,109576,15,26339


### 5.1 Merging `panchayat_inequity` with `subdist_census_2001_filtered`

The merge chain is:
1. Join `panchayat_inequity` → `PC_MIS_mapping` on `mis_index` to recover subdistrict names for each panchayat
2. Join that result → `subdist_census_2001_filtered` on `(state_name, district_name, subdistrict_name)`

**Pre-merge fix:** 31 name triplets appear more than once in `subdist_census_2001_filtered` because multiple census subdistrict IDs share the same name within a district (mostly urban Delhi subdivisions). These are aggregated by summing population columns before the merge to avoid row fan-out.

**Known data gap:** ~16,833 panchayats in `panchayat_inequity` have no match in `PC_MIS_mapping` (the mapping file only covers ~53% of panchayats), so those rows will have NaN census values.

In [57]:
name_key = ['state_name', 'district_name', 'subdistrict_name']

# Collapse duplicate name triplets in census data by summing populations
subdist_census_agg = (
    subdist_census_2001_filtered
    .groupby(name_key, dropna=False)[['tot_pop_persons', 'SC_pop_persons', 'ST_pop_persons']]
    .sum()
    .reset_index()
)

# Step 1: attach subdistrict names to panchayat_inequity via mis_index
mis_to_loc = (
    PC_MIS_mapping[['mis_index', 'State name', 'District name', 'Subdistrict name']]
    .dropna(subset=['mis_index'])
    .drop_duplicates(subset='mis_index')
    .rename(columns={
        'State name': 'state_name',
        'District name': 'district_name',
        'Subdistrict name': 'subdistrict_name'
    })
)

panchayat_with_loc = panchayat_inequity.merge(mis_to_loc, on='mis_index', how='left')

panchayat_with_loc.head()

,Unnamed: 0,mis_index,equity_score1,equity_score2,equity_score3,equity_score4,category,equity_score5,inequity_score,group,state_name,district_name,subdistrict_name
0,0,166463,0.018656,0.004735,0.094903,0.189998,3,0.084442,1.763303,SC,andhra pradesh,vizianagaram,bondapalle
1,1,166475,0.597425,-0.161602,0.302553,0.005162,2,0.015320,1.619678,SC,andhra pradesh,vizianagaram,denkada
2,2,166478,0.103851,NaN,0.120494,0.120689,3,0.022218,1.982577,SC,andhra pradesh,vizianagaram,denkada
3,3,166479,0.469600,-0.351881,0.468233,0.243940,3,0.064675,1.531420,SC,andhra pradesh,vizianagaram,denkada
4,4,166533,0.278150,0.551750,0.388139,0.166026,2,0.053694,1.289105,SC,andhra pradesh,vizianagaram,parvathipuram


In [58]:
# Step 2: attach census data on name triplet
panchayat_census = panchayat_with_loc.merge(
    subdist_census_agg, on=name_key, how='left', indicator=True
)

match_counts = panchayat_census['_merge'].value_counts()
print(f'Matched:    {match_counts.get("both", 0):,}')
print(f'Unmatched:  {match_counts.get("left_only", 0):,}')
print(f'Total rows: {len(panchayat_census):,}')

panchayat_census = panchayat_census.drop(columns='_merge')
panchayat_census.head()

Matched:    58,727
Unmatched:  12,917
Total rows: 71,644


,Unnamed: 0,mis_index,equity_score1,equity_score2,equity_score3,equity_score4,category,equity_score5,inequity_score,group,state_name,district_name,subdistrict_name,tot_pop_persons,SC_pop_persons,ST_pop_persons
0,0,166463,0.018656,0.004735,0.094903,0.189998,3,0.084442,1.763303,SC,andhra pradesh,vizianagaram,bondapalle,50531.0,5712.0,1445.0
1,1,166475,0.597425,-0.161602,0.302553,0.005162,2,0.015320,1.619678,SC,andhra pradesh,vizianagaram,denkada,46486.0,4805.0,92.0
2,2,166478,0.103851,NaN,0.120494,0.120689,3,0.022218,1.982577,SC,andhra pradesh,vizianagaram,denkada,46486.0,4805.0,92.0
3,3,166479,0.469600,-0.351881,0.468233,0.243940,3,0.064675,1.531420,SC,andhra pradesh,vizianagaram,denkada,46486.0,4805.0,92.0
4,4,166533,0.278150,0.551750,0.388139,0.166026,2,0.053694,1.289105,SC,andhra pradesh,vizianagaram,parvathipuram,58575.0,6675.0,8895.0


In [67]:
panchayat_census.to_csv('/Users/vidhi/VSCODE/Econ_191_paper/FinalPaperMaterial copy/2. Data/3. Processed/Census_Panchayat_inequity.csv', index=False)

## Dataset: Kjelsrud - NREGA

In [170]:
NREGA = pd.read_stata('/Users/vidhi/college baby/sem 6/Topics in EconRes/FinalPaperMaterial/2. Data/1. Raw/Kjelsrud/ReplicationPackage/Datafiles/nrega.dta')
NREGA.head()

,state,district2011,gp_id,year,pc_id_pre,pc_id_post,change_pc,matched_pc_post,postbank,sPostbank,...,share_lit_past,religion_frac_past,caste_frac_past,share_blocksplit,distance_km,change_ac,split_block,poverty_pre66_past,pc_dist_y,mean_share_acW
0,28.0,542,128950.0,2012.0,1.0,431.0,1.0,439.0,1627157.0,0.491032,...,0.444422,1.0,0.245299,0.0,0.738072,1.0,0.0,0.345833,2.0,1.0
1,28.0,542,128699.0,2013.0,1.0,431.0,1.0,439.0,1993932.0,0.702515,...,0.444422,1.0,0.245299,0.0,0.738072,1.0,0.0,0.345833,3.0,1.0
2,28.0,542,128694.0,2013.0,1.0,431.0,1.0,439.0,1050892.0,0.158757,...,0.444422,1.0,0.245299,0.0,0.738072,1.0,0.0,0.345833,3.0,1.0
3,28.0,542,128937.0,2013.0,1.0,431.0,1.0,439.0,1516612.0,0.427292,...,0.444422,1.0,0.245299,0.0,0.738072,1.0,0.0,0.345833,3.0,1.0
4,28.0,542,128932.0,2011.0,1.0,431.0,1.0,439.0,0.0,-0.447188,...,0.444422,1.0,0.245299,0.0,0.738072,1.0,0.0,0.345833,1.0,1.0


In [171]:
NREGA.columns

Index(['state', 'district2011', 'gp_id', 'year', 'pc_id_pre', 'pc_id_post',
       'change_pc', 'matched_pc_post', 'postbank', 'sPostbank', 'demand',
       'sDemand', 'gini_pre66_past', 'meanlog_pre66_past', 'theil_pre66_past',
       'palma_pre66_past', 'fragmentation_2004_past',
       'min_vote_share_2004_past', 'min_margin_2004_past',
       'effective_parties_2004_past', 'reservation_post', 'party_2009',
       'log_mean_pre66_past', 'pc_dist', 'cell_share_sc', 'cell_share_st',
       'cell_share_l6', 'cell_share_lit', 'cell_primary', 'cell_phc',
       'cell_phs', 'cell_tap', 'cell_power', 'cell_paved', 'primary_past',
       'phc_past', 'phs_past', 'tap_past', 'power_past', 'paved_past',
       'share_sc_past', 'share_st_past', 'share_l6_past', 'population_past',
       'urbanization_past', 'share_lit_past', 'religion_frac_past',
       'caste_frac_past', 'share_blocksplit', 'distance_km', 'change_ac',
       'split_block', 'poverty_pre66_past', 'pc_dist_y', 'mean_share_acW'],


Checking whether there are any PCs that do not change constituencies but the 'change_pc' indicator says that they did change constituencies.

In [64]:
mask = (NREGA['change_ac'] == 1) & (NREGA['pc_id_pre'] == NREGA['matched_pc_post'])
bad_rows = NREGA.loc[mask]

print("Matches found:", len(bad_rows))
display(bad_rows[['pc_id_pre', 'matched_pc_post', 'change_ac']].value_counts())

mask = (NREGA['change_ac'] == 1) & (NREGA['pc_id_pre'] == NREGA['pc_id_post'])
bad_rows = NREGA.loc[mask]

print("Matches found:", len(bad_rows))
display(bad_rows[['pc_id_pre', 'pc_id_post', 'change_ac']].value_counts())

Matches found: 294


pc_id_pre  matched_pc_post  change_ac
362.0      362.0            1.0          219
512.0      512.0            1.0           75
Name: count, dtype: int64

Matches found: 186


pc_id_pre  pc_id_post  change_ac
362.0      362.0       1.0          138
512.0      512.0       1.0           48
Name: count, dtype: int64